File: algorithms/sorting/merge-sort/MergeSort.java
import java.util.Arrays;

public final class MergeSort {
    private MergeSort() {
    }

    public static int[] mergeSort(int[] numbers) {
        if (numbers.length <= 1) return Arrays.copyOf(numbers, numbers.length);
        int middle = numbers.length / 2;
        int[] left = mergeSort(Arrays.copyOfRange(numbers, 0, middle));
        int[] right = mergeSort(Arrays.copyOfRange(numbers, middle, numbers.length));
        return merge(left, right);
    }

    private static int[] merge(int[] left, int[] right) {
        int[] output = new int[left.length + right.length];
        int leftIndex = 0;
        int rightIndex = 0;
        int outputIndex = 0;

        while (leftIndex < left.length && rightIndex < right.length) {
            if (left[leftIndex] <= right[rightIndex]) {
                output[outputIndex++] = left[leftIndex++];
            } else {
                output[outputIndex++] = right[rightIndex++];
            }
        }

        while (leftIndex < left.length) output[outputIndex++] = left[leftIndex++];
        while (rightIndex < right.length) output[outputIndex++] = right[rightIndex++];
        return output;
    }
}

File: algorithms/sorting/merge-sort/merge_sort.cpp
#include <vector>

std::vector<int> merge(const std::vector<int>& left, const std::vector<int>& right) {
    std::vector<int> output;
    output.reserve(left.size() + right.size());
    std::size_t leftIndex = 0;
    std::size_t rightIndex = 0;

    while (leftIndex < left.size() && rightIndex < right.size()) {
        if (left[leftIndex] <= right[rightIndex]) output.push_back(left[leftIndex++]);
        else output.push_back(right[rightIndex++]);
    }

    output.insert(output.end(), left.begin() + leftIndex, left.end());
    output.insert(output.end(), right.begin() + rightIndex, right.end());
    return output;
}

std::vector<int> mergeSort(const std::vector<int>& numbers) {
    if (numbers.size() <= 1) return numbers;
    const std::size_t middle = numbers.size() / 2;
    const std::vector<int> leftInput(numbers.begin(), numbers.begin() + middle);
    const std::vector<int> rightInput(numbers.begin() + middle, numbers.end());
    const std::vector<int> left = mergeSort(leftInput);
    const std::vector<int> right = mergeSort(rightInput);
    return merge(left, right);
}
